### PR TITLE
Django 1.8

### PR DESCRIPTION
**Pull request by kaedroho**
This PR adds Django 1.8 support into Wagtail

### CODE REVIEW

**Comment by kaedroho** [comment]
I think it's worth trying to get this into 0.9. Django 1.8 is scheduled for the end of this month and 1.0 might be a long wait.

**Comment by tomdyson** [comment]
I agree. How much work do you think it is? Could someone else help?

On Tuesday, 17 March 2015, Karl Hobley <notifications@github.com> wrote:

> I think it's worth trying to get this into 0.9. Django 1.8 is scheduled
> for the end of this month and 1.0 might be a long wait.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/torchbox/wagtail/pull/1060#issuecomment-82436796>.
>


-- 
sent from my phone

**Comment by kaedroho** [comment]
The work on getting Wagtail working on Django 1.8 is mostly done. Mainly waiting on the following PRs to be merged:
 - https://github.com/django/django/pull/4334
 - https://github.com/alex/django-taggit/pull/297
 - https://github.com/torchbox/django-modelcluster/pull/20
 - https://github.com/torchbox/wagtail/pull/1060

Reviews much appreciated!

File: wagtail/wagtailadmin/edit_handlers.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 from six import text_type
 
 from modelcluster.forms import ClusterForm, ClusterFormMetaclass
-
+from modelcluster.models import get_related_model
```
Comment (kaedroho) [inline wagtail/wagtailadmin/edit_handlers.py]: Would probably be best to redefine this in wagtail so an older version of modelcluster can still be used under Django 1.7

File: wagtail/wagtailadmin/tests/test_page_chooser.py [hunk 2]
```diff
@@ -104,8 +104,8 @@ def test_create_link(self):
         response = self.post({'url': 'http://www.example.com'})
         self.assertEqual(response.status_code, 200)
         self.assertContains(response, "'onload'")  # indicates success / post back to calling page
-        self.assertContains(response, "'url': 'http://www.example.com/',")
-        self.assertContains(response, "'title': 'http://www.example.com/'")
+        self.assertContains(response, "'url': 'http://www.example.com")
```
Comment (kaedroho) [inline wagtail/wagtailadmin/tests/test_page_chooser.py]: Django 1.8 no longer appends slashes to URLs without a path. I don't think this really matters but I could probably improve this test to make it use a regex or something.

File: wagtail/wagtailadmin/utils.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 from modelcluster.fields import ParentalKey
+from modelcluster.models import get_related_model
```
Comment (kaedroho) [inline wagtail/wagtailadmin/utils.py]: https://github.com/torchbox/wagtail/pull/1060/files#r26648411

File: wagtail/wagtailcore/models.py [hunk 4]
```diff
@@ -1084,6 +1084,7 @@ def as_page_object(self):
 
         # Override the possibly-outdated tree parameter fields from this revision object
         # with up-to-date values
+        obj.pk = self.page.pk
```
Comment (kaedroho) [inline wagtail/wagtailcore/models.py]: This is required as we hit Djangos unsaved instance check (outside of modelcluster as that bypasses it). Might be worth investigating...
https://docs.djangoproject.com/en/dev/releases/1.8/#assigning-unsaved-objects-to-relations-raises-an-error

File: wagtail/wagtailforms/tests.py [hunk 5]
```diff
@@ -79,29 +79,6 @@ def test_post_multiple_values(self):
         self.assertIn("baz", submission[0].form_data)
 
 
-class TestPageModes(TestCase):
```
Comment (kaedroho) [inline wagtail/wagtailforms/tests.py]: This change has been made on master: https://github.com/torchbox/wagtail/commit/af7c1de978a7647d6c2023dda43133d3523bb103

File: wagtail/wagtailusers/views/groups.py [hunk 6]
```diff
@@ -83,7 +83,8 @@ def create(request):
         Group,
         GroupPagePermission,
         formset=BaseGroupPagePermissionFormSet,
-        extra=0
+        extra=0,
+        fields=('page', 'permission_type'),
```
Comment (kaedroho) [inline wagtail/wagtailusers/views/groups.py]: Submitted separately in: https://github.com/torchbox/wagtail/pull/1059

**Comment by timheap** [comment]
I tried to combine Django 1.8 (rc1), Wagtail, and Jinja templates, and came across a problem: Jinja templates fail when the context contains a `self` key! I made a Django bug for it: https://code.djangoproject.com/ticket/24538#ticket

**Comment by kaedroho** [comment]
@timheap thanks for testing!

**Comment by huxley** [comment]
Just noticed that the Django ticket @timheap references is now marked **wontfix** because Tim has found that *self* is a special variable in Jinja2. Wondering how difficult it would be to switch Wagtail to use a different variable than *self*?

**Comment by gasman** [comment]
@huxley Yep - it sounds like our best option is to make the name 'self' switchable via a setting. Discussion is currently going on here: https://groups.google.com/d/msg/wagtail-developers/MxE6ImpVPCs/mDzyEYFZmY4J

**Comment by timheap** [comment]
I sent a pull request for Django that will help with the 'self' in templates issue: https://github.com/django/django/pull/4415

**Comment by chrxr** [comment]
Hi @timheap, please could you review this. Django 1.8 is due to be released tomorrow, and we're keen to get support into Wagtail 1.0. Thanks!

**Comment by timheap** [comment]
This looks good! I am keen to see Wagtail in Django 1.8. Wagtail 1.0 will be huge at this rate! Some thoughts on this pull request:

-----

`tox.ini` tests `py27-dj18-postgres`, but not `-mysql` or `-sqlite`, as these are commented out. Is this deliberate?

`tox.ini` contains some git URLs for dependencies still. I understand that these changes are not yet merged so this is required, just dont forget them before release!

-----

Feature detection seems to be used in many places, as opposed to version detection. In my opinion, this makes the code much harder to read and reason about, as well as hiding the true intention of the code behind comments. For example [`get_related_model`](https://github.com/torchbox/wagtail/pull/1060/files#diff-27c04bf7bf8bbfc020f4800e14363da9R1):

```python
def get_related_model(rel):
    # In Django 1.7 and under, the related model is accessed by doing: rel.model
    # This was renamed in Django 1.8 to rel.related_model. rel.model now returns
    # the base model.
    return getattr(rel, 'related_model', rel.model)
```

and [`TestRouting.test_request_serving`](https://github.com/torchbox/wagtail/pull/1060/files#diff-4cb0a1f1754ceea497976c8e1a71584fR191):

```python
# Django 1.8+
if hasattr(used_template, 'template'):
    used_template = used_template.template
```

In my opinion, checking the version is much clearer, straight forward, less error prone, and more explicit:

```python
def get_related_model(rel):
    # In Django 1.7 and under, the related model is accessed by doing: rel.model
    # This was renamed in Django 1.8 to rel.related_model.
    if DJANGO_VERSION < (1, 8):
        return rel.model
    else:
        return rel.related_model
```

```python
if DJANGO_VERSION >= (1, 8):
    used_template = used_template.template
```

This has the added benefit of being self documenting, as it is clear that the functionality differs from version to version. It also works in all cases, and is already in use where feature detection does not work (e.g. migrations involving content types: `defaults={'name': 'document'} if DJANGO_VERSION < (1, 8) else {}`). Finally, when Django 1.7 support is dropped in the future, searching the codebase for the tuple `(1, 8)` should return all instances of hacks done to support the 1.7/1.8 differences.

-----

I did not review the pull requests for django-taggit and django-modelcluster.

**Comment by kaedroho** [comment]
> tox.ini tests py27-dj18-postgres, but not -mysql or -sqlite, as these are commented out. Is this deliberate?

Yeah.. Unfortunately, the more we make Travis do the slower it gets. We're testing all the database backends on Python 3.4, but just Postgres on Python 2.7.

> tox.ini contains some git URLs for dependencies still. I understand that these changes are not yet merged so this is required, just dont forget them before release!

Maybe I'll just include them for the Django 1.8 tests for now and test Django 1.7 with released versions.

> Feature detection seems to be used in many places, as opposed to version detection. In my opinion, this makes the code much harder to read and reason about, as well as hiding the true intention of the code behind comments. 

That looks better, will change

**Comment by kaedroho** [comment]
This may require some tweaks now that the API is merged

**Comment by kaedroho** [comment]
Ready for merge. Theres still a load of deprecation warnings, I'll solve those in another PR.

**Comment by kaedroho** [comment]
Django taggit ``0.13.0`` released with my PR merged. Still waiting on ``django-modelcluster``. This PR can be merged at any time though.

**Comment by gasman** [comment]
Have cherry-picked af5cbaa, 229a666, 775da10, c2456b6, fc374bc - please rebase.

**Comment by gasman** [comment]
5f198fc, b43aff6, 55cd422, 4d8e31b all cherry-picked. 819b8cc rejected, because we're not yet ready to add 1.8 to the 'include' list (the RoutablePage failures are still outstanding as of this PR). Remaining commits to be addressed in a separate PR which I'll open shortly.

**Pull request closed without merging**